### PR TITLE
Bridge 1382

### PR DESCRIPTION
**Pull request by Erin-Mounts**
- Allow any Bridge request to be done with the background session (so it can be initiated offline) by adding a BOOL "background" parameter to the basic networking methods.
- Refactor interfaces to use NSURLSessionTask instead of NSURLSessionDataTask:
--it's the parent class of both NSURLSessionDataTask and NSURLSessionDownloadTask, so the same set of methods can be used for both with no change other than the background parameter
--it doesn't affect any calling code beyond changing the class name in some completion handler parameter lists, because NSURLSessionDataTask doesn't add any properties or methods to the parent (abstract) class
--I should have done it this way in the first place
- Have requests using the background session to access Bridge APIs respond to HTTP errors (401, 410, etc.) the same as foreground requests
- Refactor the uploadFileToBridge:etc. code to take advantage of this, and to use completion handlers in place of delegate methods
- Update SBBActivityManager updateScheduledActivities:withCompletion: to use the background session

### CODE REVIEW

File: BridgeSDK/BridgeAPI/SBBBridgeNetworkManager.m [hunk 1]
```diff
@@ -74,11 +74,12 @@ - (instancetype)initWithAuthManager:(id<SBBAuthManagerInternalProtocol>)authMana
     return self;
 }
 
-- (NSURLSessionDataTask *) doDataTask: (NSString*) method
-                            URLString: (NSString*)URLString
-                              headers: (NSDictionary *)headers
-                           parameters:(NSDictionary *)parameters
-                           completion:(SBBNetworkManagerCompletionBlock)completion
+- (NSURLSessionTask *)doDataTask:(NSString *)method
```
Comment (syoung-smallwisdom) [inline BridgeSDK/BridgeAPI/SBBBridgeNetworkManager.m]: Is this intended as an internal method only? (ie. should the old version be deprecated?)

File: BridgeSDK/BridgeAPI/SBBUploadManager.m [hunk 2]
```diff
@@ -255,6 +255,125 @@ - (SBBUploadSession *)uploadSessionForFile:(NSString *)fileURLString
     return uploadSession;
 }
 
+// When uploading a file to Bridge, first we use background download to get an SBBUploadSession object from Bridge.
+// Then we use the pre-signed URL in that SBBUploadSession to upload our file to S3.
+// When that finishes, we tell Bridge that the requested upload is completed.
+// To avoid having multiple layers of nested completion handlers in one method, they're broken out by step:
+
+// --- Here's the completion handler for the step where we tell Bridge the upload completed:
+- (void)toldBridgeFileUploadedWithTask:(NSURLSessionUploadTask *)uploadTask forBridgeSession:(SBBUploadSession *)uploadSession error:(NSError *)error
```
Comment (syoung-smallwisdom) [inline BridgeSDK/BridgeAPI/SBBUploadManager.m]: nit: I prefer "handle" to "told"... or is the meaning something different?

File: BridgeSDK/BridgeAPI/SBBBridgeNetworkManager.m (hunk 1 above)
Comment (Erin-Mounts) [inline BridgeSDK/BridgeAPI/SBBBridgeNetworkManager.m]: it's internal-only

File: BridgeSDK/BridgeAPI/SBBUploadManager.m [hunk 3]
```diff
@@ -255,6 +255,125 @@ - (SBBUploadSession *)uploadSessionForFile:(NSString *)fileURLString
...
+    if (httpStatusCode >= 300) {
+        // iOS handles redirects automatically so only e.g. 307 resource not changed etc. from the 300 range should end up here
+        // (along with all 4xx and 5xx of course)
+        NSString *description = [NSString stringWithFormat:@"Background file upload to S3 failed with HTTP status %ld", (long)httpStatusCode];
+        NSError *s3Error = [NSError errorWithDomain:SBB_ERROR_DOMAIN code:SBBErrorCodeS3UploadErrorResponse userInfo:@{NSLocalizedDescriptionKey: description}];
+        [self completeUploadOfFile:uploadTask.taskDescription withError:s3Error];
+        return;
+    }
+    
+    // tell the API we done did it
```
Comment (syoung-smallwisdom) [inline BridgeSDK/BridgeAPI/SBBUploadManager.m]: "done did it"?

File: BridgeSDK/BridgeAPI/SBBUploadManager.m (hunk 2 above)
Comment (Erin-Mounts) [inline BridgeSDK/BridgeAPI/SBBUploadManager.m]: well, it's a "handler" that gets called when we've "told" Bridge the file upload completed...

File: BridgeSDK/Networking/SBBNetworkManagerInternal.h [hunk 4]
```diff
@@ -53,9 +53,16 @@ extern NSString *kAPIPrefix;
 
 - (instancetype)initWithBaseURL:(NSString*)baseURL bridgeStudy:(NSString*)bridgeStudy;
 
-- (void)handleHTTPError:(NSError *)error task:(NSURLSessionDataTask *)task response:(id)responseObject retryObject:(APCNetworkRetryObject *)retryObject;
+- (void)handleHTTPError:(NSError *)error task:(NSURLSessionTask *)task response:(id)responseObject retryObject:(SBBNetworkRetryObject *)retryObject;
 - (NSDictionary *)headersPreparedForRetry:(NSDictionary *)headers;
 
 - (NSURL *) URLForRelativeorAbsoluteURLString: (NSString*) URLString;
 
+- (NSURLSessionTask *) doDataTask:(NSString*) method
```
Comment (syoung-smallwisdom) [inline BridgeSDK/Networking/SBBNetworkManagerInternal.h]: question answered.

File: BridgeSDK/BridgeAPI/SBBUploadManager.m (hunk 3 above)
Comment (Erin-Mounts) [inline BridgeSDK/BridgeAPI/SBBUploadManager.m]: moved this code here from where it used to be, mostly unchanged (including this comment)--probably had been reading or watching a lot of something with Southern or country dialect at the time? dunno.

File: BridgeSDK/BridgeAPI/SBBUploadManager.m [hunk 5]
```diff
@@ -255,6 +255,125 @@ - (SBBUploadSession *)uploadSessionForFile:(NSString *)fileURLString
...
+    NSString *uploadId = uploadSession.id;
+    if (!uploadId.length) {
+        NSAssert(uploadId, @"uploadId is nil");
+    }
+    NSString *ref = [NSString stringWithFormat:kSBBUploadCompleteAPIFormat, uploadId];
+    NSMutableDictionary *headers = [NSMutableDictionary dictionary];
+    [self.authManager addAuthHeaderToHeaders:headers];
+    [self.networkManager post:ref headers:headers parameters:nil background:YES completion:^(NSURLSessionTask *task, id responseObject, NSError *error) {
+        [self toldBridgeFileUploadedWithTask:uploadTask forBridgeSession:uploadSession error:error];
+    }];
```
Comment (Erin-Mounts) [inline BridgeSDK/BridgeAPI/SBBUploadManager.m]: not seeing what you're seeing here...?

**Comment by syoung-smallwisdom** [comment]
looks good

**Pull request merged**